src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs

```diff
@@ -0,0 +1,129 @@
+        private GCHandle _operationHandle = new GCHandle();
+        
+        private volatile bool _disposed = false; // To detect redundant calls.
+        
+        public WinHttpRequestState()
+        {
+            TransportContext = new WinHttpTransportContext();
+            _operationHandle = GCHandle.Alloc(this);
+        }
+
+        public static WinHttpRequestState FromIntPtr(IntPtr gcHandle)
+        {
+            var stateHandle = GCHandle.FromIntPtr(gcHandle);
+            return (WinHttpRequestState)stateHandle.Target;
+        }        
+
+        public static string UriSchemeHttps { get { return "https"; } }
+        
+        public IntPtr ToIntPtr()
+        {
+            return GCHandle.ToIntPtr(_operationHandle);
+        }
+        
+        public object Lock
+        {
+            get
+            {
+                return _lock;
+            }
+        }
+
+        public TaskCompletionSource<HttpResponseMessage> Tcs { get; set; }
+
+        public CancellationToken CancellationToken { get; set; }
+
+        public HttpRequestMessage RequestMessage { get; set; }
+
+        public WinHttpHandler Handler { get; set; }
+
+        public SafeWinHttpHandle RequestHandle { get; set; }
+
+        public Exception SavedException { get; set; }
+
+        public bool CheckCertificateRevocationList { get; set; }
+
+        public Func<
+            HttpRequestMessage,
+            X509Certificate2,
+            X509Chain,
+            SslPolicyErrors,
+            bool> ServerCertificateValidationCallback { get; set; }
+
+        public WinHttpTransportContext TransportContext { get; private set; }
+
+        public WindowsProxyUsePolicy WindowsProxyUsePolicy { get; set; }
+
+        public IWebProxy Proxy { get; set; }
+
+        public ICredentials ServerCredentials { get; set; }
+
+        public ICredentials DefaultProxyCredentials { get; set; }
+
+        public bool PreAuthenticate { get; set; }
+        
+        public HttpStatusCode LastStatusCode { get; set; }
+
+        public bool RetryRequest { get; set; }
+        
+        // Important: do not hold _lock while signaling completion of any of below TaskCompletionSources.
+        public TaskCompletionSource<bool> TcsSendRequest { get; set; }
+        public TaskCompletionSource<bool> TcsWriteToRequestStream { get; set; }
+        public TaskCompletionSource<bool> TcsInternalWriteDataToRequestStream { get; set; }
+        public TaskCompletionSource<bool> TcsReceiveResponseHeaders { get; set; }
+        public TaskCompletionSource<int> TcsReadFromResponseStream { get; set; }
+        
+        #region IDisposable Members
+        private void Dispose(bool disposing)
+        {
+            // Since there is no finalizer and this class is sealed, the disposing parameter should be TRUE.
+            Debug.Assert(disposing, "WinHttpRequestState.Dispose() should have disposing=TRUE");
+            
```

Comment:
WinHttpHandler doesn't use the same SafeHandle pattern as WebSockets.Client (which does the dispose of handles when the State object is disposed).